### DISCUSSION TITLE
How does work (CR)Touch leveling?

### DISCUSSION THREAD

**Discussion by AlexeyPechnikov** [mriscoc/Ender3V2S1#608] [Q&A]
It looks over complicated how the firmware processes (CR) Touch sensors. If I understand correct it provides raw measurements via serial interface (a mesh) and we need to calculate a bed surface model using the measures. Also, the firmware requires z-levels to be entered manually… does it mean residuals between the measured values and the model or should we enter all the serial reading?! I don’t see any way to obtain the measures with high accuracy without the sensor…actually, we have the sensor exactly for this task but the firmware asks for all the values instead of the sensor usage. Also, we need to adjust g-codes outside of the firmware to follow the bed level changes using the model. Ok, that’s not a problem to calculate but why do not use just a switch like to “apply CR touch” to do all the calculations inside the firmware without any mesh creations and the model calculations and z-offset manual input? I didn’t check the codes jet but I checked  the release logs and menu items (instead of missed documentation) trying to understand how the firmware works.

**Comment by Mysteoa** [comment]
The way I understand it is that CR will probe the bed and record the depth relative to itself. It will create a mesh and adjust the median, so most points are as close to 0. Then you have to add M420 gcode to enable bed leveling mesh at the beginning of your gcode, so it gets used during the print. 

The thing you describe is I think for creating a Manual mesh using the CR touch of the data. What CR is usually used is with UBL FW that can automatically create the mesh.

**Comment by mriscoc** [comment]
Currently the firmware uses the UBL from Marlin, relevant links with information is here: https://github.com/mriscoc/Ender3V2S1/wiki/UBL-versions

**Comment by AlexeyPechnikov** [comment]
Ok, I see my question is not clear so let me explain. How to get diagnostic output like to g29 & m48 v4 but for UBL?

**Reply by mriscoc** [reply]
ABL `G29` is in UBL `G29 P1`
`M48 V4` is working as usual
